aten/src/ATen/record_function.h

```diff
@@ -174,75 +187,91 @@ struct TORCH_API RecordFunction {
       F fn,
       std::vector<c10::IValue>&& args,
       int64_t current_sequence_nr = -1) {
-    inputs_ = std::move(args);
+    if (!isActive()) {
+      return;
+    }
+    state_->inputs_ = std::move(args);
     before(fn, current_sequence_nr);
   }
 
-  // Calls end callbacks
+  // Calls end callbacks. After end(), accessors will no longer provide useful results.
   void end();
 
   inline RecordFunctionHandle handle() const {
-    return handle_;
+    TORCH_CHECK(state_, "Called handle() on inactive RecordFunction");
```

Comment:
we saw in the past that TORCH_CHECK in inline functions can be costly (as it generates a full blown if statement and can throw the compiler's inline logic off the tracks). I'd try to replace it with TORCH_INTERNAL_ASSERT_DEBUG_ONLY or replace it all together to see if you get further savings

btw, TORCH_CHECK -> TORCH_INTERNAL_ASSERT anyway. They are kind of similar but generate different error messages: TORCH_CHECK is more of a user error, INTERNAL_ASSERT is for PyTorch invariant violations (which can happen only because of code bugs)